components/script/dom/range.rs

```diff
@@ -893,6 +895,52 @@ impl RangeMethods for Range {
         // Step 6.
         s
     }
+
+    // https://dvcs.w3.org/hg/innerhtml/raw-file/tip/index.html#extensions-to-the-range-interface
+    fn CreateContextualFragment(&self, fragment: DOMString) -> Fallible<Root<DocumentFragment>> {
+        // Step 1.
+        let node = self.StartContainer();
+        let element = match node.type_id() {
+            NodeTypeId::Document(_) | NodeTypeId::DocumentFragment => None,
+            NodeTypeId::Element(_) => Some(node),
+            NodeTypeId::CharacterData(CharacterDataTypeId::Comment) |
+            NodeTypeId::CharacterData(CharacterDataTypeId::Text) => node.GetParentNode(),
+            NodeTypeId::CharacterData(CharacterDataTypeId::ProcessingInstruction) |
+            NodeTypeId::DocumentType => unreachable!(),
+        };
+
+        // Step 2.
+        let elem_copy = element.clone().unwrap();
+        let document = elem_copy.downcast::<Document>().unwrap();
+        let element = if element.is_none() {
+            Element::new(Atom::from("body"),
+                         ns!(html),
+                         None,
+                         &document)
+        } else if document.is_html_document() {
+            let c_elem = element.unwrap();
+            let elem = c_elem.downcast::<Element>().unwrap();
+            if elem.local_name() == "html" && *elem.namespace() == ns!(html) {
+                Element::new(Atom::from("body"),
+                             ns!(html),
+                             None,
+                             &document)
+            } else {
+                elem.root_element()
+            }
+        } else {
+            element.unwrap().downcast::<Element>().unwrap().root_element()
+        };
+
+        // Step 3.
+        let fragment_node = element.upcast::<Node>().parse_fragment(fragment);
```

Comment:
It seems strange that I have to upcast to get back a node in order to use this method.